worker/interactions.go

```diff
@@ -35,61 +61,71 @@ func interactionMiddleware(ir InteractionRecorder, routes map[string]jape.Handle
+		bus:    w.bus,
+		logger: w.logger,
 
-func (w *worker) RecordHostScan(scans ...hostdb.HostScan) {
-	w.interactionsMu.Lock()
-	defer w.interactionsMu.Unlock()
+		flushCtx:      w.shutdownCtx,
+		flushInterval: flushInterval,
 
-	w.interactionsScans = append(w.interactionsScans, scans...)
-	w.tryFlushInteractionsBuffer()
+		hostScans:         make([]hostdb.HostScan, 0),
+		priceTableUpdates: make([]hostdb.PriceTableUpdate, 0),
+	}
 }
 
-func (w *worker) RecordPriceTableUpdate(ptUpdates ...hostdb.PriceTableUpdate) {
-	w.interactionsMu.Lock()
-	defer w.interactionsMu.Unlock()
+func (r *hostInteractionRecorder) RecordHostScan(scans ...hostdb.HostScan) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	r.hostScans = append(r.hostScans, scans...)
+	if r.interactionsFlushTimer == nil {
+		r.interactionsFlushTimer = time.AfterFunc(r.flushInterval, r.flush)
+	}
+}
 
-	w.interactionsPriceTableUpdates = append(w.interactionsPriceTableUpdates, ptUpdates...)
-	w.tryFlushInteractionsBuffer()
+func (r *hostInteractionRecorder) RecordPriceTableUpdate(ptUpdates ...hostdb.PriceTableUpdate) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	r.priceTableUpdates = append(r.priceTableUpdates, ptUpdates...)
+	if r.interactionsFlushTimer == nil {
+		r.interactionsFlushTimer = time.AfterFunc(r.flushInterval, r.flush)
+	}
 }
 
-func (w *worker) tryFlushInteractionsBuffer() {
-	// If a thread was scheduled to flush the buffer we are done.
-	if w.interactionsFlushTimer != nil {
-		return
+func (r *hostInteractionRecorder) Stop(ctx context.Context) {
+	r.mu.Lock()
+	if r.interactionsFlushTimer != nil {
+		r.interactionsFlushTimer.Stop()
 	}
+	r.flushCtx = ctx
+	r.flushInterval = 0
+	r.mu.Unlock()
 
-	// Otherwise we schedule a flush.
-	w.interactionsFlushTimer = time.AfterFunc(w.busFlushInterval, func() {
-		w.interactionsMu.Lock()
-		w.flushInteractions()
-		w.interactionsMu.Unlock()
-	})
+	r.flush()
 }
 
-// flushInteractions flushes the worker's interaction buffer to the bus.
-func (w *worker) flushInteractions() {
-	if len(w.interactionsScans) > 0 {
-		if err := w.bus.RecordHostScans(w.shutdownCtx, w.interactionsScans); err != nil {
-			w.logger.Errorw(fmt.Sprintf("failed to record scans: %v", err))
-		} else {
-			w.interactionsScans = nil
+func (r *hostInteractionRecorder) flush() {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	if len(r.hostScans) > 0 {
+		if err := r.bus.RecordHostScans(r.flushCtx, r.hostScans); err != nil {
+			r.logger.Errorw(fmt.Sprintf("failed to record scans: %v", err))
+		} else if err == nil {
+			r.hostScans = nil
 		}
 	}
-	if len(w.interactionsPriceTableUpdates) > 0 {
-		if err := w.bus.RecordPriceTables(w.shutdownCtx, w.interactionsPriceTableUpdates); err != nil {
-			w.logger.Errorw(fmt.Sprintf("failed to record price table updates: %v", err))
-		} else {
-			w.interactionsPriceTableUpdates = nil
+	if len(r.priceTableUpdates) > 0 {
+		if err := r.bus.RecordPriceTables(r.flushCtx, r.priceTableUpdates); err != nil {
```

Comment:
This `r.flushCtx` is the worker's `shutdownCtx` right? So a shutdown will abort this, produce an error in the logs and then `Stop` will succeed. 
To the user it will still feel like something sometimes fails on shutdown.

Reply:
on shutdown we swap out the recorder's context with the one that we call shutdown with

Reply:
So to give a little more context (😏 ):

The recorders are a little tricky because we use a timer to schedule flushes, this leads to an annoying race condition on worker shutdown where we close the shutdown context immediately and then want to still record host interactions from downloads and/or uploads that happen when we are winding down.

My solution is to rename `shutdownCtx` to `flushCtx`, so it's clear it's the context we use to flush interactions to the bus. Then, on shutdown, we swap out the `flushCtx` and replace it with the context we use to shut down the worker with (this is different from the shutdown ctx).

We can do that because on flush we now check whether the current flush context is cancelled, if so we don't flush. This, in combination with logging an error if we were unable to flush the remaining interactions on stop, ticks all the boxes. We essentially allow recorders to flush for as long as the context passed it on shutdown is not cancelled, we don't log errors when we don't need to and we make sure to log an error when we fail to flush interactions.

Reply:
Sry if I wasn't clear enough with my question. I get that you are swapping it out but you only swap it out after `shutdownCtx` is closed right?
So if we are in the middle of a `flush` call, the flush will be interrupted with an error, then we swap the context and then we flush again successfully.

The reason this works is because the `select` checking the `flushCtx` isn't thread safe since `flushCtx` might be cancelled outside of `mu`. Only swapping the context is protected by `mu`. 
So you might end up with `flushCtx` getting closed right after the `select` but before `RecordPriceTables`.

Reply:
I thought about that but you can't get around with that (I don't think). Not without interacting with the recorder before cancelling the worker shutdown. The way it's now is that you will indeed get an error, but the records will (or might) still make it into the bus because we flush them on stop.

Reply:
I no longer like the shutdown context on the worker by the way. Especially when you want to shut down gracefully it's merely a trigger to indicate we're shutting down but other than that it's useless because you still need _the actual_ context with which we are shutting down the worker gracefully.